### PR TITLE
trivial: make fwupd-tests remote mandatory for test devices

### PR DESCRIPTION
**Pull request by superm1**
Debian (currently) packages fwupd-tests as a separate package from fwupd.  If test mode is enabled a lot of test things work, but locally installed metadata specifically doesn't.

Make it clearer when test mode is enabled that the fwupd-tests remote also needs to be present.

Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [x] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

**Comment by hughsie** [comment]
Won't that cause the daemon just not work? Shouldn't we show something (the error?) to the user when they try to enable test mode?

**Comment by superm1** [comment]
Ooohh good point!  I like that approach much better.

**Comment by superm1** [comment]
> Ooohh good point! I like that approach much better.

I did push those changes, can you take another look?

File: src/fu-engine.c [hunk 1]
```diff
@@ -797,8 +821,19 @@ fu_engine_modify_config(FuEngine *self,
...
+		/* many options need a reboot after this */
+		if (!fu_config_set_value(FU_CONFIG(self->config), section, key, value, error))
+			return FALSE;
+
+		/* reload remotes */
+		if (g_strcmp0(key, "TestDevices") == 0) {
+			g_object_unref(self->remote_list);
+			self->remote_list = fu_remote_list_new();
+			if (!fu_engine_load_remotes(self, error))
+				return FALSE;
```
Comment (hughsie) [inline src/fu-engine.c:833]: this chunk feels like it's going to cause problems; all FwupdRemote's in the first list are now ~invalid. Can we add the flag, invalidate the "loaded" boolean and call fu_engine_load_remotes() to get the new list perhaps?

Comment (superm1) [reply src/fu-engine.c]: Fixed

File: src/fu-remote-list.c [hunk 2]
```diff
@@ -658,6 +658,25 @@ fu_remote_list_load_metainfos(XbBuilder *builder, GError **error)
 	return TRUE;
 }
 
+gboolean
+fu_remote_list_toggle_test_remotes(FuRemoteList *self, gboolean enable, GError **error)
```
Comment (hughsie) [inline src/fu-remote-list.c:662]: something like fu_remote_list_set_testing_remote_enabled() would be better -- toggle changes true->false and false->true.

File: src/fu-remote-list.c [hunk 3]
```diff
@@ -658,6 +658,25 @@ fu_remote_list_load_metainfos(XbBuilder *builder, GError **error)
...
+{
+	g_return_val_if_fail(FU_IS_REMOTE_LIST(self), FALSE);
+
+	/* not yet initialized */
+	if (self->silo == NULL)
+		return TRUE;
+
+	self->testing_remote = enable;
```
Comment (hughsie) [inline src/fu-remote-list.c:670]: maybe check if `self->testing_remote == enable` and return here?

File: src/fu-tool.c [hunk 4]
```diff
@@ -2121,37 +2121,68 @@ fu_util_remote_enable(FuUtilPrivate *priv, gchar **values, GError **error)
...
 {
 	if (!fu_util_start_engine(priv, FU_ENGINE_LOAD_FLAG_NONE, priv->progress, error))
 		return FALSE;
 
-	if (!fu_engine_modify_config(priv->engine,
-				     "fwupd",
-				     "TestDevices",
-				     enable ? "true" : "false",
-				     error))
+	if (!fu_util_toggle_test_devices(priv, FALSE, error))
```
Comment (hughsie) [inline src/fu-tool.c:2138]: fu_util_set_test_devices_enabled perhaps?

**Comment by hughsie** [comment]
> How do you feel about bringing enable-test-devices and disable-test-devices commands to fwupdmgr

Ngggggg, that's a lot of boilerplate and overhead IHO.

File: src/fu-remote-list.c (hunk 3 above)
Comment (superm1) [reply src/fu-remote-list.c:673]: fixed

**Comment by superm1** [comment]
> > How do you feel about bringing enable-test-devices and disable-test-devices commands to fwupdmgr
> 
> Ngggggg, that's a lot of boilerplate and overhead IHO.

Ack

**Pull request merged**